File: app/src/main/java/org/adaway/db/dao/HostEntryDao.java
package org.adaway.db.dao;

import androidx.annotation.Nullable;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;

import org.adaway.db.entity.HostEntry;
import org.adaway.db.entity.HostListItem;
import org.adaway.db.entity.ListType;

import java.util.List;
import java.util.regex.Pattern;

import static androidx.room.OnConflictStrategy.REPLACE;
import static org.adaway.db.entity.ListType.REDIRECTED;

/**
 * This interface is the DAO for {@link HostEntry} records.
 *
 * @author Bruce BUJON (bruce.bujon(at)gmail(dot)com)
 */
@Dao
public interface HostEntryDao {
    Pattern ANY_CHAR_PATTERN = Pattern.compile("\\*");
    Pattern A_CHAR_PATTERN = Pattern.compile("\\?");

    @Query("DELETE FROM `host_entries`")
    void clear();

    @Query("INSERT INTO `host_entries` SELECT DISTINCT `host`, `type`, `redirection` FROM `hosts_lists` WHERE `type` = 0 AND `enabled` = 1")
    void importBlocked();

    @Query("SELECT host FROM hosts_lists WHERE type = 1 AND enabled = 1")
    List<String> getEnabledAllowedHosts();

    @Query("DELETE FROM `host_entries` WHERE `host` LIKE :hostPattern")
    void allowHost(String hostPattern);

    @Query("SELECT * FROM hosts_lists WHERE type = 2 AND enabled = 1 ORDER BY host ASC, source_id DESC")
    List<HostListItem> getEnabledRedirectedHosts();

    @Insert(onConflict = REPLACE)
    void redirectHost(HostEntry redirection);

    /**
     * Synchronize the host entries based on the current hosts lists table records.
     */
    default void sync() {
        clear();
        importBlocked();
        for (String allowedHost : getEnabledAllowedHosts()) {
            allowedHost = ANY_CHAR_PATTERN.matcher(allowedHost).replaceAll("%");
            allowedHost = A_CHAR_PATTERN.matcher(allowedHost).replaceAll("_");
            allowHost(allowedHost);
        }
        for (HostListItem redirectedHost : getEnabledRedirectedHosts()) {
            HostEntry entry = new HostEntry();
            entry.setHost(redirectedHost.getHost());
            entry.setType(REDIRECTED);
            entry.setRedirection(redirectedHost.getRedirection());
            redirectHost(entry);
        }
    }

    @Query("SELECT * FROM `host_entries` ORDER BY `host`")
    List<HostEntry> getAll();

    @Query("SELECT `type` FROM `host_entries` WHERE `host` == :host LIMIT 1")
    ListType getTypeOfHost(String host);

    @Query("SELECT IFNULL((SELECT `type` FROM `host_entries` WHERE `host` == :host LIMIT 1), 1)")
    ListType getTypeForHost(String host);

    @Nullable
    @Query("SELECT * FROM `host_entries` WHERE `host` == :host LIMIT 1")
    HostEntry getEntry(String host);
}

File: app/src/main/java/org/adaway/db/dao/HostEntryDao.kt
package org.adaway.db.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import org.adaway.db.entity.HostEntry
import org.adaway.db.entity.HostListItem
import org.adaway.db.entity.ListType
import org.adaway.db.entity.ListType.REDIRECTED
import java.util.regex.Pattern

@Dao
interface HostEntryDao {
    @Query("DELETE FROM `host_entries`")
    fun clear()

    @Query("INSERT INTO `host_entries` SELECT DISTINCT `host`, `type`, `redirection` FROM `hosts_lists` WHERE `type` = 0 AND `enabled` = 1")
    fun importBlocked()

    @get:Query("SELECT host FROM hosts_lists WHERE type = 1 AND enabled = 1")
    val enabledAllowedHosts: List<String>

    @Query("DELETE FROM `host_entries` WHERE `host` LIKE :hostPattern")
    fun allowHost(hostPattern: String)

    @get:Query("SELECT * FROM hosts_lists WHERE type = 2 AND enabled = 1 ORDER BY host ASC, source_id DESC")
    val enabledRedirectedHosts: List<HostListItem>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun redirectHost(redirection: HostEntry)

    fun sync() {
        clear()
        importBlocked()
        for (allowedHost in enabledAllowedHosts) {
            val hostPattern = A_CHAR_PATTERN.matcher(
                ANY_CHAR_PATTERN.matcher(allowedHost).replaceAll("%")
            ).replaceAll("_")
            allowHost(hostPattern)
        }
        for (redirectedHost in enabledRedirectedHosts) {
            val entry = HostEntry().apply {
                host = redirectedHost.host
                type = REDIRECTED
                redirection = redirectedHost.redirection
            }
            redirectHost(entry)
        }
    }

    @get:Query("SELECT * FROM `host_entries` ORDER BY `host`")
    val all: List<HostEntry>

    @Query("SELECT `type` FROM `host_entries` WHERE `host` == :host LIMIT 1")
    fun getTypeOfHost(host: String): ListType

    @Query("SELECT IFNULL((SELECT `type` FROM `host_entries` WHERE `host` == :host LIMIT 1), 1)")
    fun getTypeForHost(host: String): ListType

    @Query("SELECT * FROM `host_entries` WHERE `host` == :host LIMIT 1")
    fun getEntry(host: String): HostEntry?

    companion object {
        private val ANY_CHAR_PATTERN: Pattern = Pattern.compile("\\*")
        private val A_CHAR_PATTERN: Pattern = Pattern.compile("\\?")
    }
}
